File: leetcode/0155. Min Stack.py
'''
Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
getMin() -- Retrieve the minimum element in the stack.
Example:
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin();   --> Returns -3.
minStack.pop();
minStack.top();      --> Returns 0.
minStack.getMin();   --> Returns -2
'''
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack=[]
        self.minstack=[]

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.stack.append(x)
        if not self.minstack:
            self.minstack.append(x)
        else:
            if x<=self.minstack[-1]:
                self.minstack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        s=self.stack.pop()
        if s==self.minstack[-1]:
            self.minstack.pop()

    def top(self):
        """
        :rtype: int
        """
        if not self.stack:
            return None
        return self.stack[-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        if not self.minstack:
            return None
        return self.minstack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: archives/C#/0155. Min Stack.cs
public class MinStack {
    
    Stack<int> stackS=new Stack<int>();
    Stack<int> minStack=new Stack<int>();
    /** initialize your data structure here. */
    public MinStack() {
        
    }
    
    public void Push(int x) {
        stackS.Push(x);
        if (minStack.Count==0 || minStack.Peek()>=x){
            minStack.Push(x);
        }
    }
    
    public void Pop() {
        int tmp=stackS.Pop();
        if(tmp==minStack.Peek()){
            minStack.Pop();
        }
    }
    
    public int Top() {
        return stackS.Peek();
    }
    
    public int GetMin() {
         return minStack.Peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(x);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */
